beater/telemetry.go

```diff
@@ -0,0 +1,118 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"github.com/elastic/apm-server/beater/config"
+	"github.com/elastic/apm-server/idxmgmt"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/idxmgmt/ilm"
+	"github.com/elastic/beats/v7/libbeat/monitoring"
+)
+
+var apmRegistry = monitoring.GetNamespace("state").GetRegistry().NewRegistry("apm-server")
+var configRegistry = apmRegistry.NewRegistry("configuration")
+
+type configTelemetry struct {
+	rumEnabled                *monitoring.Int
+	apiKeysEnabled            *monitoring.Int
+	kibanaEnabled             *monitoring.Int
+	pipelinesEnabled          *monitoring.Int
+	pipelinesOverwrite        *monitoring.Int
+	setupTemplateEnabled      *monitoring.Int
+	setupTemplateOverwrite    *monitoring.Int
+	setupTemplateAppendFields *monitoring.Int
+	ilmEnabled                *monitoring.Int
+	ilmSetupEnabled           *monitoring.Int
+	ilmSetupOverwrite         *monitoring.Int
+	ilmSetupRequirePolicy     *monitoring.Int
+	jaegerGRPCEnabled         *monitoring.Int
+	jaegerHTTPEnabled         *monitoring.Int
+	sslEnabled                *monitoring.Int
```

Comment:
What's the reason for making these ints, and not bools? In libbeat, the `management.enabled` field is a bool: https://github.com/elastic/beats/blob/7ec32e4d2e1dcd9dcfb9b61e49fff794deeb2c2a/libbeat/cmd/instance/beat.go#L322

Reply:
I don't think that *exactly one* use case sets any precedence :)
It is because ints might be easier to aggregate than bools, and give more granular information. 
IIUC, if a cluster deploys eg. 10 apm-servers, with ints we can see how many of those have rum enabled, but with bools we can only answer "is there at least one rum enabled?" or "do all have rum enabled?" questions.
Even if it turns out we only want to answer at-least-one questions, I don't see any drawbacks to ints - they just give more flexibility.

Reply:
> I don't think that exactly one use case sets any precedence :)

In libbeat this is the only occurrence, but there are plenty of others in the telemetry schema: https://github.com/elastic/kibana/blob/master/x-pack/plugins/telemetry_collection_xpack/schema/xpack_plugins.json

(There are a couple of fields called "enabled" and "disabled" with type `long`; those are referring to number of rules enabled/disabled.)

In the event that there's just one other occurrence (for this or other changes), I would prefer that we do not take a second approach unless there's a very good reason to.

> It is because ints might be easier to aggregate than bools, and give more granular information.
>
> IIUC, if a cluster deploys eg. 10 apm-servers, with ints we can see how many of those have rum enabled, but with bools we can only answer "is there at least one rum enabled?" or "do all have rum enabled?" questions.

We can use a filter aggregation for this, with doc_count giving the same answer.

Reply:
> I would prefer that we do not take a second approach unless there's a very good reason to

I don't think using one or other type for a variable qualifies as a "second approach" - ints instead of bools is not like breaking the molds precisely. Still the reason was anyways easier aggregation.

Similarly, regarding your other (related) comments, I imagined that if we start adding more data, and index it like:
```
{ state: {
    configuration: {
        rum_enabled: 1,
        apikeys_enabled: 1,
        ...
   },
   events: {
   ...
   },
   agents: {
   ...
   },
}
```    

it will be more readable than an endless bag of:
```
{state: {
    key: {
        nested_key: {
             nested_key: value
        },
    }
{
```

without any order or structure. 
   

I don't find that to be less consistent (there isn't anything documented/prescribed I am ware of), but more trying to do what fits best.

Just answering questions, thou. I am not strongly opinionated about and will go trough suggestions